subprojects/core/src/integTest/groovy/org/gradle/api/tasks/DeferredTaskDefinitionIntegrationTest.groovy

```diff
@@ -251,4 +252,43 @@ class DeferredTaskDefinitionIntegrationTest extends AbstractIntegrationSpec {
         outputContains("Received :task1")
         result.assertNotOutput("task2")
     }
+
+    @Issue("https://github.com/gradle/gradle/issues/5148")
+    def "can get a task by name with a filtered collection"() {
+        buildFile <<'''
+            tasks.createLater("task1", SomeTask) {
+                println "Configure ${path}"
+            }
+            
+            tasks.create("other") {
+                dependsOn tasks.withType(SomeTask).getByName("task1")
+            }
+        '''
+
+        when:
+        run "other"
+
+        then:
+        outputContains("Create :task1")
+    }
+
+    def "fails to get a task by name when it does not match the collection filter"() {
```

Comment:
perhaps also a test case to show that `withType(x).getByName(y)` fails when there is no element with type x and name y (or perhaps add an issue if it doesn't work yet).

Reply:
Yeah, it works - I've added a test for it.